### PR TITLE
Implement GraqlEntityProjectionMaker

### PR DESCRIPTION
**Pull request by hellohanchen**
Resolves https://github.com/yahoo/elide/issues/890 and https://github.com/yahoo/elide/issues/823

## Description
Implement `GraphQLEntityProjectionMaker` to construct an `EntityProjection` from a GraphQL `Document` object
- Implement `FragmentResolver` to parse and validate fragments in GraphQL request
- 
- Update `Entity` class in Elide/GraphQL module to contain projection in which this entity is fetched, this change match the latest API in `PersistentResource`.
- Update `GraphQLEndpoint` to use `GraphQLEntityProjectionMaker` for parsing incoming GraphQL request, this change is also applied for test classes.
- Update `PersistentResourceFetcher` to use projection
- Add enum `Keyword` for all GraphQL keywords we use in this module

## How Has This Been Tested?
These changes pass all unit test in `elide-graphql` module.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/GraphQLEntityProjectionMaker.java [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+                .map(definition -> (FragmentDefinition) definition)
+                .collect(Collectors.toList());
+        fragmentMap.putAll(FragmentResolver.resolve(fragmentDefinitions));
+
+        // resolve operation definitions
+        parsedDocument.getDefinitions().forEach(definition -> {
+            if (definition instanceof OperationDefinition) {
+                // Operations would be converted into EntityProjection tree
+                OperationDefinition operationDefinition = (OperationDefinition) definition;
+                if (operationDefinition.getOperation() == OperationDefinition.Operation.MUTATION) {
```
Comment (aklish) [inline elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/GraphQLEntityProjectionMaker.java]: I don't think we want to skip mutations.  We just want to skip subscriptions.

File: elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/GraphQLEntityProjectionMaker.java [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        Attribute argumentAttribute = entityProjection.getAttributeByName(argumentName);
+        com.yahoo.elide.request.Argument elideArgument = com.yahoo.elide.request.Argument.builder()
+                .name(argumentName)
+                .value(argument.getValue())
+                .build();
+
+        if (argumentAttribute == null) {
+            argumentAttribute = Attribute.builder()
+                    .type(entityDictionary.getType(entityType, argumentName))
+                    .name(argumentName)
```
Comment (aklish) [inline elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/GraphQLEntityProjectionMaker.java]: I think this is wrong.  The name of the attribute should be the field name - not the argument name.

File: elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/GraphQLEntityProjectionMaker.java [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+     */
+    private void addField(Field field, EntityProjection parentProjection) {
+        Class<?> parentType = parentProjection.getType();
+        String fieldName = field.getName();
+
+        // this field would either be a relationship field or an attribute field
+        if (entityDictionary.getRelationshipType(parentType, fieldName) != RelationshipType.NONE) {
+            // handle the case of a relationship field
+            final Class<?> relationshipType =
+                    entityDictionary.getParameterizedType(parentType, fieldName) == null
```
Comment (aklish) [inline elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/GraphQLEntityProjectionMaker.java]: This check for null shouldn't be necessary.  getParameterized type will return the correct type regardless of whether the type is actually parameterized.

File: elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/GraphQLEntityProjectionMaker.java (hunk 2 above)
Comment (hellohanchen) [reply elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/GraphQLEntityProjectionMaker.java]: I think the argument here is something like `id`, which is one of the fields of the projected entity, isn't it?

**Pull request closed without merging**